### PR TITLE
[#1058] Serially deploy modules that share a cachedir

### PR DESCRIPTION
**Pull request by justinstoller**
When we raised the default threadpool for deploying modules above 1, we
began to see issues when caching a remote locally if that remote/cache
was shared by multiple modules within the Puppetfile.

To resolve this, when we create the module we group it with any other
module that may share its cachedir. We then pull modules by cachedir off
of the queue and serially process those.

Please add all notable changes to the "Unreleased" section of the CHANGELOG.

### CODE REVIEW

**Review by Magisus — APPROVED** [review]
This looks like it'll do the trick, while maintaining as much parallelism as is safe.

**Comment by Magisus** [comment]
I heard plans to add an acceptance test, which I think we should do. Was that going to be added to this PR? Or should we add this in, use the [existing test](https://github.com/puppetlabs/r10k/blob/master/integration/tests/purging/content_not_purged_at_root.rb) that failed because of this as-is to verify it, and add the new one later (in the interest of the release timeline)?

File: spec/unit/puppetfile_spec.rb
Comment (mwaggett) [inline spec/unit/puppetfile_spec.rb:331]: is this intentionally still 'module4'? or should it be 5?

File: spec/unit/puppetfile_spec.rb
Comment (mwaggett) [inline spec/unit/puppetfile_spec.rb:341]: Can this be `length.times.map` instead of hard-coding 4?

**Comment by mwaggett** [comment]
If we're in a rush, I'm okay using the existing test, but I do think we should ultimately have one that's explicitly testing this. I'll leave it up to @justinstoller to decide whether it can be done quick enough to be included here.

**Comment by adrienthebo** [comment]
I was under the mistaken assumption that serializing access to caches had been resolved as part of the parallel modules pull request. Given that this is still an issue we may want to consider dropping the default `pool_size` back to 1 while we verify that parallelization is safe. Part of this concern is that we have the [`WithModules`](https://github.com/puppetlabs/r10k/blob/master/lib/r10k/environment/with_modules.rb#L55-L64) environment that has another entrypoint into interacting with modules that this pull request might not account for.

File: lib/r10k/puppetfile.rb
Comment (adrienthebo) [inline lib/r10k/puppetfile.rb:192]: As this method deals with properties of modules, we can push this detail down into the modules themselves instead of probing objects for methods and acting from there.

File: lib/r10k/puppetfile.rb
Comment (adrienthebo) [inline lib/r10k/puppetfile.rb:237]: This might have some interesting interactions with https://github.com/puppetlabs/r10k/blob/master/lib/r10k/environment/with_modules.rb#L55-L64.

**Comment by justinstoller** [comment]
How about this? https://github.com/puppetlabs/r10k/pull/1060

**Comment by justinstoller** [comment]
Ugh, it looks like the `WithModules` implementation maybe just copied and pasted code from the `Puppetfile` implementation? We probably should refactor out that shared code.

File: lib/r10k/puppetfile.rb
Comment (dhollinger) [reply lib/r10k/puppetfile.rb:237]: I would feel safer if this was addressed before the PR is merged. Just to avoid any potential unforeseen issues.

**Review by dhollinger — COMMENTED** [review]
I don't really have anything additional to add that @adrienthebo and @mwaggett haven't already pointed out. I would like to see the potential interaction with `WithModules` to be solved before merging though

**Comment by justinstoller** [comment]
Yeah, sorry, I didn't mean to ghost, y'all. When we thought we could make this in before a deadline we gave it a go, and when it became apparent we weren't going to make it for that deadline it dropped way down our priority list.

File: lib/r10k/puppetfile.rb
Comment (Magisus) [reply lib/r10k/puppetfile.rb:237]: I may be missing something, but I'm not sure how this would matter in `WithModules`. That class doesn't have a `concurrent_accept` mode so far as I can tell, so its modules would be processed first, always in serial, and then the Puppetfile would be processed and this logic would handle the potential conflict.

I can't find any other other places that respect `pool_size` besides what's in the Puppetfile here. What did y'all think was potentially problematic?

**Comment by Magisus** [comment]
I don't appear to be able to push to this branch. Is that something @justinstoller can change, or do I need to open a new PR to pick this up?

**Comment by justinstoller** [comment]
I pushed this code to this branch so anyone can take it over: https://github.com/puppetlabs/r10k/tree/GH-1058_dont-concurrently-modify-same-cache

File: spec/unit/puppetfile_spec.rb
Comment (Magisus) [reply spec/unit/puppetfile_spec.rb:331]: I think it doesn't really matter, since calling `spy` will still create unique objects regardless of that string. But I'll update it.

**Comment by Magisus** [comment]
Superseded by https://github.com/puppetlabs/r10k/pull/1081

**Pull request closed without merging**
